File: Problem_1.js
// Time Complexity : All methods have O(1)
// Space Complexity : O(n) space in worst case
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach
//Reused the hash set technique from Design-1 class to design the following hash map. Instead of storing boolean val, store the actual value provided by user. 

var MyHashMap = function() {
    this.storage = new Array(1000).fill(null);
    this.slots = 1000;
    this.slotItem = 1000;
};

MyHashMap.prototype.slotIdx = function(key) {
    return key%this.slots;
};

MyHashMap.prototype.slotItemIdx = function(key) {
    return parseInt(key/this.slotItem);
};

/** 
 * @param {number} key 
 * @param {number} value
 * @return {void}
 */
MyHashMap.prototype.put = function(key, value) {
    let slot = this.slotIdx(key);
    let slotItem = this.slotItemIdx(key);
    if(this.storage[slot] == null){
        this.storage[slot] = new Array(this.slotItem).fill(null);
    }
    this.storage[slot][slotItem] = value;
};

/** 
 * @param {number} key
 * @return {number}
 */
MyHashMap.prototype.get = function(key) {
    let s = this.slotIdx(key);
    let sI = this.slotItemIdx(key);
    if(this.storage[s] == null || this.storage[s][sI] == null)
        return -1;
    return this.storage[s][sI]; 
};

/** 
 * @param {number} key
 * @return {void}
 */
MyHashMap.prototype.remove = function(key) {
    let s = this.slotIdx(key);
    let sI = this.slotItemIdx(key);
    if(this.storage[s] == null)
        return;
    this.storage[s][sI] = null; 
};

/** 
 * Your MyHashMap object will be instantiated and called as such:
 * var obj = new MyHashMap()
 * obj.put(key,value)
 * var param_2 = obj.get(key)
 * obj.remove(key)
 */

File: Problem1.java
// Time Complexity : Amortized O(1)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
// S30 ID: Sawan Chakraborty_RN69DEC2024

import java.util.Stack;

class MyQueue {
    Stack<Integer> in;// input stack
    Stack<Integer> out;// output stack acting as queue
    public MyQueue() {
        this.in = new Stack<>();
        this.out = new Stack<>();
    }
    
    public void push(int x) {
        this.in.push(x);
    }
    
    public int pop() {
        peek(); // calling peek to move elements to out stack
        return this.out.pop();
    }
    
    public int peek() {
        if(this.out.isEmpty()){
            while(!this.in.isEmpty()){
                this.out.push(this.in.pop()); // pushing elements to out stack
            }
        }
        return this.out.peek();
    }
    
    public boolean empty() {
        return this.in.isEmpty() && this.out.isEmpty(); // checking empty condition
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */